### ISSUE TITLE
get serialization to be configurable with redis backend, others ?

### ISSUE THREAD

**Issue by sqlalchemy-bot** [sqlalchemy/dogpile.cache#18]
**Migrated issue, originally created by Michael Bayer ([zzzeek](https://github.com/zzzeek))**

**Comment by zzzeek** [comment]
another year later.    do we still want to be able to pass in loads/dumps callables?    I guess putting them on a Serializer class is, while a little more J2EE than I am these days, more consistent with how other thigns in dogpile have been done.     the serilaizer gets the CachedValue and your serializer does with it what it wants, throws it away, adds it to your JSON or whatever, doesn't matter.

**Comment by jvanasco** [comment]
I think it is still an important feature to allow a choice of serializers. `PIckle` is a great default, but being locked into it isn't so great.  Being able to store things in `msgpack` made one `Redis` instance about 25% more efficient for me (and some other tricks made it even more efficient!)

Design wise... there are a lot of ways that I would prefer to implement this, but I think the most backwards compatible would be changing the regions to accept a `loads`/`dumps` args that default to `None`, pass those onto the backend, and let the backend fall back to `Pickle` if nothing is configured.

**Comment by zzzeek** [comment]
OK yes, this is what I favored throughout this discussion as I re-read it, I believe also moving to github makes these kinds of things easier to navigate and read as having to go to other repos on bitbucket is basically something I had a negative emotional reaction towards, I apologize.       Pull requests that illustrate code changes are best, but i guess in this case we just need to implement, right?    cc @4383

**Comment by jvanasco** [comment]
> as having to go to other repos on bitbucket is basically something I had a negative emotional reaction towards

IMHO that is how everyone who doesn't work for Bitbucket feels when using their system.  Sorry for overloading you with so many options, trying to give you an easier choice was really just complicating things.

Generally speaking though...  the backends would accept loads/dumps like this:

https://github.com/jvanasco/dogpile_backend_redis_advanced/blob/master/dogpile_backend_redis_advanced/cache/backends/redis_advanced.py#L155-L156

then just call `self.loads()` and `self.dumps()`

Looking at the `Region` code in dogpile, I don't think that would actually need to change; the backend arguments aren't filtered or anything.

**Comment by zzzeek** [comment]
yeah it's not a hard change but they just have to deal with the CachedValue thing explicitly.   They can of course not store that info or store it somewhere else or use some kind of range based approach , that's all about the serialization strategy.

**Comment by zzzeek** [comment]
we can include a basic JSON plugin which will make some basic JSON of the CachedValue then call loads() / dumps() on the value they give us.    Folks can use that as a starting point for other kinds of json / serialization or whatever.   a msgpack plugin also, sure.

**Comment by jvanasco** [comment]
IMHO, the easiest way to handle that is to bundle the (de)serialization into the backend's `loads`/`dumps` functions.

These example in a public repo is msgpack, but I do the same with json in private projects.  

msgpack - `dumps` is fine as/is, but `loads` needs to wrap the tuple into a `CachedValue`

https://github.com/jvanasco/dogpile_backend_redis_advanced/blob/master/tests/cache/test_redis_backend.py#L252-L272

This example will amuse and possibly anger you... The CachedValue's metadata is tossed before serialization. When reading it out, it is recreated with the current time so the value never expires as far as dogpile is concerned.

https://github.com/jvanasco/dogpile_backend_redis_advanced/blob/master/tests/cache/test_redis_backend.py#L252-L272
